master/pkg/stream/stream.go

```diff
@@ -109,15 +116,15 @@ func (p *Publisher[T]) Broadcast(events []Event[T]) {
 
 	// check each event against each subscription
 	for _, sub := range p.Subscriptions {
-		func(){
+		func() {
 			for _, ev := range events {
-				var msg *websocket.PreparedMessage
+				var msg interface{}
 				if ev.After != nil && sub.filter(*ev.After) {
 					// update, insert, or fallin: send the record to the client.
-					msg = (*ev.After).UpsertMsg()
+					msg = (*ev.After).UpsertMsg(sub.UpsertFunc)
 				} else if ev.Before != nil && sub.filter(*ev.Before) {
 					// deletion or fallout: tell the client the record is deleted.
-					msg = (*ev.Before).DeleteMsg()
+					msg = (*ev.Before).DeleteMsg(sub.DeleteFunc)
```

Comment:
I am wondering how we could make this UpsertFunc/DeleteFunc feel less awkward.  I think what you have written is roughly what I described, but now that I see it I don't really care for it.

But I see something useful: this Broadcast codepath is the only path where caching is meaningful.  The CollectStartupMessages path or whatever doesn't need to cache because it pulls TrialMsgs out of the database, streams them to just one client, and discards them.  Caching isn't useful there.

So that leads me to believe that we can rearrange where the caching happens.  Instead of having the caching in each message type, like `TrialMsg.upsertCache`, you could do the caching on the `stream.Event` object:

```go
type Event[T Msg] struct {
    Before *T `json:"before"`
    After  *T `json:"after"`
    
    upsertCache interface{}
    deleteCache interface{}
}
```

This makes way more sense than what I prototyped.  Before, we only ever used `ev.After.upsertCache` and `ev.Before.deleteCache`, so even though there were four possible cached messages per message, only two were in use. Now that will be explicit.  Also, now the only time that caching _can_ happen is in Broadcast(), which matches what _was_ happening.  Which is always nice.

I think we can also get rid of the concept of "override" functions and the corresponding `if func == nil` checks everywhere.  Instead of passing in an optional prepare overrides, you can pass in mandatory prepare functions.  And I don't think it has to be a member of the `Message` interface anymore.

There's still something awkward that I'm having trouble putting my finger on though.  Something about the outer `"trial":` or `"trials_deleted":` key that gets put on every message.  Maybe if there were a couple more types in `pkg/stream/stream.go`:

```go
type UpsertMsg struct {
    Type string
    Msg Msg
}

type DeleteMsg struct {
    Type string
    Keys string  // range-encoded string
}
```

If you had those, then I think you could create UpsertMsg and DeletedMsg directly in the Broadcast loop, and pass the result into a single Prepare function (no more prepare overrides, and no more separate upsert prepare vs delete prepare):

```go
var msg interface{}
if ev.After != nil && sub.filter(*ev.After) {
  if ev.upsertCache == nil {
    ev.upsertCache = prepareFn(UpsertMsg(msg.UpsertTypeStr(), *ev.After))
  }
  msg = ev.upsertCache
} else if ev.Before != nil && sub.filter(*ev.Before) {
  if ev.deleteCache == nil {
    ev.deleteCache = prepareFn(DeleteMsg(msg.DeleteTypeStr(), *ev.Before))
  }
  msg = ev.deleteCache
} else {
  continue
}
```

Note that I invented `Msg.UpsertTypeStr()` and `Msg.DeleteTypeStr()` to make that work, but I think it make more sense.

Maybe a `NewUpsertMsg(msg)` constructor could automatically read `msg.UpsertTypeStr()`, which you could use in the `internal/stream/trial.go` code too.

Reply:
Can you spend some time on this idea, and see if there's a way to make things simpler than what we have today?